### PR TITLE
Feature/Listen for responses and capability changes in Isdl

### PR DESCRIPTION
**Pull request by bilal-alsharifi**
This PR is **ready** for review.

### Risk
This PR makes **minor** API changes.

### Summary

- Add `addOnRPCResponseListener` to `Isdl` interface because some managers need to listen to responses for some specific RPCs.
- Add `addOnSystemCapabilityListener` to `SystemCapabilityManager` and `Isdl` to listen for capability changes
- Add a generic `OnReponseListener`
- Add `addOnHmiStatusListener` and `removeOnHmiStatusListener` methods to `SdlManager`

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: sdl_android/src/main/java/com/smartdevicelink/proxy/SdlProxyBase.java [hunk 1]
```diff
@@ -51,6 +51,7 @@
 import com.smartdevicelink.SdlConnection.ISdlConnectionListener;
 import com.smartdevicelink.SdlConnection.SdlConnection;
 import com.smartdevicelink.SdlConnection.SdlSession;
+import com.smartdevicelink.api.ProxyBridge;
```
Comment (joeygrover) [inline sdl_android/src/main/java/com/smartdevicelink/proxy/SdlProxyBase.java]: This import should be removed

File: sdl_android/src/main/java/com/smartdevicelink/proxy/SystemCapabilityManager.java [hunk 2]
```diff
@@ -47,7 +49,28 @@ public void parseRAIResponse(RegisterAppInterfaceResponse response){
...
+	 */
+	private void notifyListeners(SystemCapabilityType systemCapabilityType, Object capability) {
+		synchronized(LISTENER_LOCK){
+			CopyOnWriteArrayList<OnSystemCapabilityListener> listeners = onSystemCapabilityListeners.get(systemCapabilityType);
+			if(listeners != null && listeners.size() > 0) {
+				for (OnSystemCapabilityListener listener : listeners) {
+					if (capability != null) {
+						listener.onCapabilityRetrieved(capability);
+					} else {
+						listener.onError(String.format("capability: %s is null", systemCapabilityType.toString()));
```
Comment (joeygrover) [inline sdl_android/src/main/java/com/smartdevicelink/proxy/SystemCapabilityManager.java]: I'm not sure if this is really an error case. Some systems might not support all capabilities and some are not mandatory parameters so they might be null.

Comment (bilal-alsharifi) [reply sdl_android/src/main/java/com/smartdevicelink/proxy/SystemCapabilityManager.java]: Right, that makes sense. I will fix that.

File: sdl_android/src/main/java/com/smartdevicelink/proxy/rpc/listeners/OnRPCResponseListener.java [hunk 3]
```diff
@@ -1,77 +1,51 @@
...
+
+public abstract class OnRPCResponseListener extends OnRPCListener {
+
+
+	public OnRPCResponseListener(){
+		setListenerType(OnRPCListener.UPDATE_LISTENER_TYPE_BASE_RPC);
+	}
+
+	public final void onReceived(final RPCMessage message){
+		//Do nothing
```
Comment (joeygrover) [inline sdl_android/src/main/java/com/smartdevicelink/proxy/rpc/listeners/OnRPCResponseListener.java]: This should actually check if the message is a response and call the `onResponse` method

File: sdl_android/src/main/java/com/smartdevicelink/api/ProxyBridge.java [hunk 4]
```diff
@@ -123,7 +127,10 @@ public boolean onRPCReceived(final RPCMessage message){
 			CopyOnWriteArrayList<OnRPCListener> listeners = rpcListeners.get(id);
 			if(listeners!=null && listeners.size()>0) {
 				for (OnRPCListener listener : listeners) {
-					listener.onRpcReceived(id,message);
+					if (message instanceof RPCResponse && listener instanceof OnRPCResponseListener){
+						RPCResponse response = (RPCResponse) message;
+						((OnRPCResponseListener)listener).onResponse(response.getCorrelationID(), response);
```
Comment (joeygrover) [inline sdl_android/src/main/java/com/smartdevicelink/api/ProxyBridge.java]: Is this necessary? I believe this should simply be the `OnRPCListener` and no modifications to this method.

File: sdl_android/src/main/java/com/smartdevicelink/api/SdlManager.java [hunk 5]
```diff
@@ -636,6 +653,16 @@ public boolean removeOnRPCNotificationListener(FunctionID notificationId, OnRPCN
 			return proxy.removeOnRPCNotificationListener(notificationId,listener);
 		}
 
+		@Override
+		public void addOnRPCResponseListener(final FunctionID responseId, final OnRPCResponseListener listener) {
```
Comment (joeygrover) [inline sdl_android/src/main/java/com/smartdevicelink/api/SdlManager.java]: This doesn't make sense. The response listener is specific to a correlation id sent by the request. This should only be `addOnRPCListener` and take a `OnRPCListener`. The remove method should be modified accordingly.

File: sdl_android/src/main/java/com/smartdevicelink/proxy/interfaces/ISdl.java [hunk 6]
```diff
@@ -149,6 +150,20 @@
      */
     boolean removeOnRPCNotificationListener(FunctionID notificationId, OnRPCNotificationListener listener);
 
+    /**
+     * Add an OnRPCResponseListener for specified response
+     * @param responseId FunctionID of the response that is to be listened for
+     * @param listener listener that should be added for the response ID
+     */
+    void addOnRPCResponseListener(FunctionID responseId, OnRPCResponseListener listener);
```
Comment (joeygrover) [inline sdl_android/src/main/java/com/smartdevicelink/proxy/interfaces/ISdl.java]: This doesn't make sense. The response listener is specific to a correlation id sent by the request. This should only be addOnRPCListener and take a OnRPCListener. The remove method should be modified accordingly.

**Comment by bilal-alsharifi** [comment]
@theRealBrett modifications look good. Thanks.

**Pull request merged**